subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/LazyTaskExecution.java

```diff
@@ -0,0 +1,205 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.ImmutableSortedSet;
+import org.gradle.api.internal.OverlappingOutputs;
+import org.gradle.internal.id.UniqueId;
+
+import java.util.Map;
+
+public class LazyTaskExecution extends TaskExecution {
+    private ImmutableSortedMap<String, Long> inputFilesSnapshotIds;
+    private ImmutableSortedMap<String, Long> outputFilesSnapshotIds;
+    private Long discoveredFilesSnapshotId;
+    private final FileSnapshotRepository snapshotRepository;
+    private ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot;
+    private ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot;
+    private FileCollectionSnapshot discoveredFilesSnapshot;
+
+    /**
+     * Creates a mutable copy of the given snapshot.
+     */
+    public LazyTaskExecution(FileSnapshotRepository snapshotRepository, TaskExecutionSnapshot taskExecutionSnapshot) {
+        this(
+            snapshotRepository,
+            taskExecutionSnapshot.getBuildInvocationId(),
+            taskExecutionSnapshot.getTaskImplementation(),
+            taskExecutionSnapshot.getTaskActionsImplementations(),
+            taskExecutionSnapshot.getInputProperties(),
+            taskExecutionSnapshot.getCacheableOutputProperties(),
+            taskExecutionSnapshot.getDeclaredOutputFilePaths(),
+            taskExecutionSnapshot.isSuccessful(),
+            null
+        );
+        this.inputFilesSnapshotIds = taskExecutionSnapshot.getInputFilesSnapshotIds();
+        this.discoveredFilesSnapshotId = taskExecutionSnapshot.getDiscoveredFilesSnapshotId();
+        this.outputFilesSnapshotIds = taskExecutionSnapshot.getOutputFilesSnapshotIds();
+    }
+
+    public LazyTaskExecution(
+        FileSnapshotRepository snapshotRepository,
+        UniqueId buildInvocationId,
+        ImplementationSnapshot taskImplementation,
+        ImmutableList<ImplementationSnapshot> taskActionsImplementations,
+        ImmutableSortedMap<String, ValueSnapshot> inputProperties,
+        ImmutableSortedSet<String> outputPropertyNames,
+        ImmutableSet<String> declaredOutputFilePaths,
+        ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot,
+        FileCollectionSnapshot discoveredFilesSnapshot,
+        ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot,
+        OverlappingOutputs detectedOverlappingOutputs
+    ) {
+        this(
+            snapshotRepository,
+            buildInvocationId,
+            taskImplementation,
+            taskActionsImplementations,
+            inputProperties,
+            outputPropertyNames,
+            declaredOutputFilePaths,
+            false,
```

Comment:
I am not convinced we should pass `false` in here. If the outcome is `UP-TO-DATE`, then the current execution would be different to the previous execution, since the current execution actually was not successful, while the previous one was. WDYT?